### DISCUSSION TITLE
Goofy Script De Dup question

### DISCUSSION THREAD

**Discussion by wesux** [Chimildic/goofy#215] [Принцип работы]
In the description of 'DeDup Tracks' it says it de dups by ID and NAME. Is there anyway to deDup only by ID?

**Comment by wesux** [comment]
Or how would I use the 'offsetDuration' argument?

this is what I have:

`function updateNewRelease(){
    let recentTracks = RecentTracks.get(900);
    let newReleaseTracks = Source.getTracks([
        
        { name: 'Current 300', id: '1JYzRcJVrd7zMTz7SFS4BM' },
        
    ]);

     Filter.dedupTracks(newReleaseTracks);
     Filter.rangeTracks(newReleaseTracks, {
       
        
    }); 
    
    
    Selector.keepLast(newReleaseTracks, 50);
        
    Playlist.saveWithReplace({
        // id: 'вашеId',
        name: '100TEST',
        tracks: newReleaseTracks,
        randomCover: 'update',
    });
}`

However, this is removing songs with the same name but they are not the same song with same ID.

**Reply by Chimildic** [reply]
Why you need to tracks with the same names?

```js
Filter.dedupTracks(newReleaseTracks, 0); // 0 ms
Filter.dedupTracks(newReleaseTracks, 10000); // 10 seconds in ms
```

**Reply by wesux** [reply]
SOme songs ahve the same title but are not the same song.

**Reply by wesux** [reply]
I just want to remove songs with the same ID.

**Reply by wesux** [reply]
Unfortunately 

```
`Filter.dedupTracks(newReleaseTracks, 0); // 0 ms
Filter.dedupTracks(newReleaseTracks, 10000); // 10 seconds in ms`
```

these did not work.

**Comment by wesux** [comment]
Unless I am using this incorrectly, It looks to be that there are a couple of problems with the 'dedup' filter.

1. If it finds duplicate songs with matching ID and Name it removes BOTH of them from the playlist. (should keep 1)
2. If it finds 2 DIFFERENT songs with the same title it removes BOTH songs from the playlist. (should keep both, as they are completely different songs)

**Reply by Chimildic** [reply]
> couple of problems with the 'dedup'

Have you real example or it just you unchecked theory?

> Some songs ahve the same title but are not the same song.

The title to check includes artist name. That why _skillet hero_ and _faouzia hero_ are different by name.

**Reply by wesux** [reply]
My appologies. I am new to this. This thing you have made is amazing!
Here is the problem that I am running into.

My original playlist contains duplicate songs AND songs with the same title.

EXAMPLE 1:
the song: "Strawberry Chainsaw" by "JAWNY" is in the playlist  'Current 300' 2 times.
Running the 'dedup' filter removes both instances of the song so it appears 0 times in the new playlist '100TEST'.

EXAMPLE 2:
The songs: "Tonight" by "Phoenix, Ezra Koening" and "Tonight" by "Peter McPoland" appear in the playlist 'Current 300'.
Running the 'dedup' filter removes both of these songs and neither one appears in the new playlist '100TEST'.

Should I be using the 'dedup' filter some other way than the following:

     function updateNewRelease(){
     let recentTracks = RecentTracks.get(900);
     let newReleaseTracks = Source.getTracks([

         { name: 'Current 300', id: '1JYzRcJVrd7zMTz7SFS4BM' },
    
     ]);

      Filter.dedupTracks(newReleaseTracks);
      Filter.rangeTracks(newReleaseTracks, {
   
    
     }); 


     Selector.keepLast(newReleaseTracks, 50);
    
     Playlist.saveWithReplace({
         // id: 'вашеId',
         name: '100TEST',
         tracks: newReleaseTracks,
         randomCover: 'update',
     });
     }

**Reply by Chimildic** [reply]
I think that dedup function has nothing to do with it. Remove line with `Selector`.

**Reply by wesux** [reply]
The dedup function is absolutely causing this to happen. If I remove the dedup filter, the songs all appear in the playlist.

EXAMPLE 1: (without dedup filter)
the song: "Strawberry Chainsaw" by "JAWNY" appears 2 times in the playlist '100TEST'

EXAMPLE 2: (without dedup filter)
The songs: "Tonight" by "Phoenix, Ezra Koening" and "Tonight" by "Peter McPoland" also get sent to '100TEST'.

If i add the dedep filter NONE of the above songs get included in the '100TEST' playlist.

The line with selector is only there to keep the playlist at 50 songs.

**Reply by Chimildic** [reply]
You want to dedup playlist with 300 tracks, but request result, after dedup, only last 50 tracks. Such selection removed other right tracks.
Run code without `Selector.keepLast` and watch result. In order to round count of tracks you need to use `keepFirst`.

**Reply by wesux** [reply]
You are correct. Taking the selector out works. But how do I make it so the new playlist '100TEST' has only 50 songs?

**Reply by Chimildic** [reply]
Just use `Selector.keepFirst` or `Selector.keepRandom`. And remember that in such case lost some tracks.
